src/services/component.ts

```diff
@@ -13,83 +13,75 @@ export type ComponentProps<TComponent extends Component> = TComponent extends Co
-          return ''
-        }
-
-        if (isPromise(values)) {
-          // @ts-expect-error there isn't a way to check if suspense is used in the component without accessing a private property
-          if (instance?.suspense) {
-            return h(suspenseAsyncPropsWrapper(component, values))
-          }
-
-          return h(asyncPropsWrapper(component, values))
-        }
-
-        return h(component, values)
-      }
-    },
-  })
-}
 
-/**
- * Creates a component wrapper that binds async props which does not require suspense
- */
-function asyncPropsWrapper<TComponent extends Component>(component: TComponent, props: Promise<ComponentProps<TComponent>>): Component {
-  return defineComponent({
-    name: 'AsyncPropsWrapper',
-    expose: [],
-    setup() {
-      const values = ref()
-
-      ;(async () => {
-        values.value = await props
-      })()
-
-      return () => {
-        if (values.value instanceof Error) {
-          return ''
-        }
-
-        if (values.value) {
-          return h(component, values.value)
-        }
-
-        return ''
+export const ComponentPropsWrapper = defineComponent((input: { component: Component, props: unknown }) => {
+  const instance = getCurrentInstance()
+
+  return () => {
+    if (input.props instanceof Error) {
+      return ''
+    }
+
+    if (isPromise(input.props)) {
+      // @ts-expect-error there isn't a way to check if suspense is used in the component without accessing a private property
+      if (instance?.suspense) {
+        return h(SuspenseAsyncComponentPropsWrapper, { component: input.component, props: input.props })
       }
-    },
-  })
-}
 
-/**
- * Creates a component wrapper that binds async props which requires suspense
- */
-function suspenseAsyncPropsWrapper<TComponent extends Component>(component: TComponent, props: Promise<ComponentProps<TComponent>>): Component {
-  return defineComponent({
-    name: 'SuspenseAsyncPropsWrapper',
-    expose: [],
-    async setup() {
-      const values = await props
-
-      return () => h(component, values)
-    },
-  })
-}
+      return h(AsyncComponentPropsWrapper, { component: input.component, props: input.props })
+    }
+
+    return h(input.component, input.props)
+  }
+}, {
+  props: ['component', 'props'],
+})
+
+const AsyncComponentPropsWrapper = defineComponent((input: { component: Component, props: unknown }) => {
+  const values = ref()
+
+  watch(() => input.props, async (props) => {
+    values.value = await props
+  }, { immediate: true, deep: true })
+
+  return () => {
+    if (values.value instanceof Error) {
+      return ''
+    }
+
+    if (values.value) {
+      return h(input.component, values.value)
+    }
+
+    return ''
+  }
+}, {
+  props: ['component', 'props'],
+})
+
+const SuspenseAsyncComponentPropsWrapper = defineComponent(async (input: { component: Component, props: unknown }) => {
+  const values = ref()
+
+  values.value = await input.props
```

Comment:
So the suspense version is different because it uses top-level await for props, right?

And we only use this version when we know that the current instance has a Vue Suspense component above?